File: pipeline/reimpl/cobol_decimal_op.py
"""
CobolDecimal operation dispatcher — parity test fixture, not a real reimpl.

This module is NOT a reimplementation of any COBOL program. It exists to
exercise `cobol_decimal.CobolDecimal` from the same JSON I/O contract used
by the differential harness, so the Java side's `CobolDecimal.java` can
be tested for byte-for-byte equivalence on inputs that the W1 unit tests
don't randomize over (overflow paths, divide buffer edge cases, etc.).

The Java twin lives at:
  pipeline/reimpl/java/runner/.../programs/CobolDecimalOp.java
and is registered in ProgramRegistry as "COBOL_DECIMAL_OP".

Operation grammar:
  op == "set"
    inputs:  digits, scale, signed, usage, on_size_error, value, rounded
    outputs: value, error
  op == "add" | "subtract" | "multiply" | "divide"
    inputs:  a_digits, a_scale, a_signed, a_value,
             b_digits, b_scale, b_signed, b_value
    outputs: result_value, result_digits, result_scale, error
  op == "assign_to"
    inputs:  src_digits, src_scale, src_signed, src_value,
             target_digits, target_scale, target_signed, rounded
    outputs: value, error
  op == "from_display"
    inputs:  digits, scale, signed, raw
    outputs: value, error
  op == "storage_bytes"
    inputs:  digits, scale, signed, usage
    outputs: storage_bytes, error

Numeric values (`value`, `a_value`, `result_value`, etc.) are passed as
their canonical Decimal string forms — both languages format them via the
same convention so byte-for-byte equality is the gate.
"""

import sys
from decimal import Decimal, InvalidOperation
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from cobol_decimal import CobolDecimal, CobolOverflowError


def _str_value(cd: CobolDecimal) -> str:
    """Canonical decimal string for a CobolDecimal value.

    Forces the stored Decimal through `str()` which gives the same shape
    Java's `BigDecimal.toPlainString()` produces. Both languages MUST
    agree on this format.
    """
    return str(cd.value)


def _parse_bool(raw) -> bool:
    if isinstance(raw, bool):
        return raw
    s = str(raw).strip().lower()
    return s in ("true", "1", "yes", "y")


def _parse_int(raw, fallback=0) -> int:
    if raw is None or raw == "":
        return fallback
    return int(str(raw).strip())


def run_vector(inputs: dict) -> dict:
    """Single-operation dispatcher. See module docstring for op grammar."""
    op = str(inputs.get("op", "")).strip().lower()

    try:
        if op == "set":
            digits = _parse_int(inputs.get("digits"), 9)
            scale = _parse_int(inputs.get("scale"), 0)
            signed = _parse_bool(inputs.get("signed", "true"))
            usage = str(inputs.get("usage", "DISPLAY"))
            ose = str(inputs.get("on_size_error", "truncate"))
            value = str(inputs.get("value", "0"))
            rounded = _parse_bool(inputs.get("rounded", "false"))

            cd = CobolDecimal(digits=digits, scale=scale, signed=signed,
                              usage=usage, on_size_error=ose)
            cd.set(Decimal(value), rounded=rounded)
            return {"value": _str_value(cd), "error": ""}

        if op in ("add", "subtract", "multiply", "divide"):
            a = CobolDecimal(
                digits=_parse_int(inputs.get("a_digits"), 9),
                scale=_parse_int(inputs.get("a_scale"), 0),
                signed=_parse_bool(inputs.get("a_signed", "true")),
            )
            a.set(Decimal(str(inputs.get("a_value", "0"))))

            b = CobolDecimal(
                digits=_parse_int(inputs.get("b_digits"), 9),
                scale=_parse_int(inputs.get("b_scale"), 0),
                signed=_parse_bool(inputs.get("b_signed", "true")),
            )
            b.set(Decimal(str(inputs.get("b_value", "0"))))

            if op == "add":
                result = a.add(b)
            elif op == "subtract":
                result = a.subtract(b)
            elif op == "multiply":
                result = a.multiply(b)
            else:  # divide
                result = a.divide(b)

            return {
                "result_value": _str_value(result),
                "result_digits": str(result.digits),
                "result_scale": str(result.scale),
                "error": "",
            }

        if op == "assign_to":
            src = CobolDecimal(
                digits=_parse_int(inputs.get("src_digits"), 9),
                scale=_parse_int(inputs.get("src_scale"), 0),
                signed=_parse_bool(inputs.get("src_signed", "true")),
            )
            src.set(Decimal(str(inputs.get("src_value", "0"))))
            target = CobolDecimal(
                digits=_parse_int(inputs.get("target_digits"), 9),
                scale=_parse_int(inputs.get("target_scale"), 0),
                signed=_parse_bool(inputs.get("target_signed", "true")),
            )
            rounded = _parse_bool(inputs.get("rounded", "false"))
            src.assign_to(target, rounded=rounded)
            return {"value": _str_value(target), "error": ""}

        if op == "from_display":
            cd = CobolDecimal(
                digits=_parse_int(inputs.get("digits"), 9),
                scale=_parse_int(inputs.get("scale"), 0),
                signed=_parse_bool(inputs.get("signed", "true")),
            )
            raw = inputs.get("raw")
            cd.from_display(raw)
            return {"value": _str_value(cd), "error": ""}

        if op == "storage_bytes":
            cd = CobolDecimal(
                digits=_parse_int(inputs.get("digits"), 9),
                scale=_parse_int(inputs.get("scale"), 0),
                signed=_parse_bool(inputs.get("signed", "true")),
                usage=str(inputs.get("usage", "DISPLAY")),
            )
            return {"storage_bytes": str(cd.storage_bytes), "error": ""}

        return {"error": f"unknown op: {op!r}"}

    except CobolOverflowError as e:
        return {"value": "", "error": f"OVERFLOW: {e}"}
    except (InvalidOperation, ValueError, ZeroDivisionError) as e:
        return {"value": "", "error": f"{type(e).__name__}: {e}"}

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/CobolDecimalOp.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.cobol.CobolDecimal;
import com.modernization.masquerade.cobol.CobolOverflowError;
import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Parity test fixture, NOT a real reimplementation. Mirrors
 * {@code pipeline/reimpl/cobol_decimal_op.py}.
 *
 * <p>Exposes {@link CobolDecimal} operations through the differential harness
 * JSON contract so the Python and Java sides can be tested for byte-for-byte
 * equivalence on inputs that the W1 unit tests don't cover (overflow paths,
 * divide buffer edge cases, randomized fuzzing).
 *
 * <p>See the Python module's docstring for the operation grammar — every
 * Java code path here mirrors the Python one-for-one.
 */
public class CobolDecimalOp implements ProgramRunner {

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String op = inputs.getOrDefault("op", "").trim().toLowerCase();
        Map<String, String> out = new LinkedHashMap<>();

        try {
            switch (op) {
                case "set":         return doSet(inputs);
                case "add":         return doBinary(inputs, "add");
                case "subtract":    return doBinary(inputs, "subtract");
                case "multiply":    return doBinary(inputs, "multiply");
                case "divide":      return doBinary(inputs, "divide");
                case "assign_to":   return doAssignTo(inputs);
                case "from_display": return doFromDisplay(inputs);
                case "storage_bytes": return doStorageBytes(inputs);
                default:
                    out.put("error", "unknown op: '" + op + "'");
                    return out;
            }
        } catch (CobolOverflowError e) {
            out.put("value", "");
            out.put("error", "OVERFLOW: " + e.getMessage());
            return out;
        } catch (ArithmeticException e) {
            out.put("value", "");
            out.put("error", "ZeroDivisionError: " + e.getMessage());
            return out;
        } catch (NumberFormatException e) {
            out.put("value", "");
            out.put("error", "InvalidOperation: " + e.getMessage());
            return out;
        } catch (Exception e) {
            out.put("value", "");
            out.put("error", e.getClass().getSimpleName() + ": " + e.getMessage());
            return out;
        }
    }

    // ── op handlers ──────────────────────────────────────────────────────

    private Map<String, String> doSet(Map<String, String> in) {
        int digits = parseInt(in.get("digits"), 9);
        int scale = parseInt(in.get("scale"), 0);
        boolean signed = parseBool(in.get("signed"), true);
        String usage = in.getOrDefault("usage", "DISPLAY");
        String oseStr = in.getOrDefault("on_size_error", "truncate");
        CobolDecimal.OnSizeError ose = "raise".equalsIgnoreCase(oseStr)
                ? CobolDecimal.OnSizeError.RAISE : CobolDecimal.OnSizeError.TRUNCATE;
        String value = in.getOrDefault("value", "0");
        boolean rounded = parseBool(in.get("rounded"), false);

        CobolDecimal cd = new CobolDecimal(digits, scale, signed, usage, ose);
        cd.set(new BigDecimal(value), rounded);

        Map<String, String> out = new LinkedHashMap<>();
        out.put("value", strValue(cd));
        out.put("error", "");
        return out;
    }

    private Map<String, String> doBinary(Map<String, String> in, String op) {
        CobolDecimal a = new CobolDecimal(
                parseInt(in.get("a_digits"), 9),
                parseInt(in.get("a_scale"), 0),
                parseBool(in.get("a_signed"), true)
        );
        a.set(new BigDecimal(in.getOrDefault("a_value", "0")));

        CobolDecimal b = new CobolDecimal(
                parseInt(in.get("b_digits"), 9),
                parseInt(in.get("b_scale"), 0),
                parseBool(in.get("b_signed"), true)
        );
        b.set(new BigDecimal(in.getOrDefault("b_value", "0")));

        CobolDecimal result;
        switch (op) {
            case "add":      result = a.add(b); break;
            case "subtract": result = a.subtract(b); break;
            case "multiply": result = a.multiply(b); break;
            case "divide":   result = a.divide(b); break;
            default: throw new IllegalStateException(op);
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("result_value", strValue(result));
        out.put("result_digits", String.valueOf(result.digits()));
        out.put("result_scale", String.valueOf(result.scale()));
        out.put("error", "");
        return out;
    }

    private Map<String, String> doAssignTo(Map<String, String> in) {
        CobolDecimal src = new CobolDecimal(
                parseInt(in.get("src_digits"), 9),
                parseInt(in.get("src_scale"), 0),
                parseBool(in.get("src_signed"), true)
        );
        src.set(new BigDecimal(in.getOrDefault("src_value", "0")));

        CobolDecimal target = new CobolDecimal(
                parseInt(in.get("target_digits"), 9),
                parseInt(in.get("target_scale"), 0),
                parseBool(in.get("target_signed"), true)
        );
        boolean rounded = parseBool(in.get("rounded"), false);
        src.assignTo(target, rounded);

        Map<String, String> out = new LinkedHashMap<>();
        out.put("value", strValue(target));
        out.put("error", "");
        return out;
    }

    private Map<String, String> doFromDisplay(Map<String, String> in) {
        CobolDecimal cd = new CobolDecimal(
                parseInt(in.get("digits"), 9),
                parseInt(in.get("scale"), 0),
                parseBool(in.get("signed"), true)
        );
        // The "raw" input may legitimately be a digit string, "SPACES", or empty.
        // Java's runner contract serializes everything as string, including null
        // → we treat the empty string as null since the Python side handles
        // both identically (both → 0).
        Object raw = in.containsKey("raw") ? in.get("raw") : null;
        cd.fromDisplay(raw);

        Map<String, String> out = new LinkedHashMap<>();
        out.put("value", strValue(cd));
        out.put("error", "");
        return out;
    }

    private Map<String, String> doStorageBytes(Map<String, String> in) {
        CobolDecimal cd = new CobolDecimal(
                parseInt(in.get("digits"), 9),
                parseInt(in.get("scale"), 0),
                parseBool(in.get("signed"), true),
                in.getOrDefault("usage", "DISPLAY")
        );
        Map<String, String> out = new LinkedHashMap<>();
        out.put("storage_bytes", String.valueOf(cd.storageBytes()));
        out.put("error", "");
        return out;
    }

    // ── helpers ──────────────────────────────────────────────────────────

    /**
     * Canonical decimal string for a CobolDecimal value. MUST match Python's
     * {@code str(decimal_value)} convention byte-for-byte: trailing zeros
     * preserved, no exponential notation, leading sign on negatives only.
     */
    private static String strValue(CobolDecimal cd) {
        return cd.value().toPlainString();
    }

    private static int parseInt(String s, int fallback) {
        if (s == null || s.isEmpty()) return fallback;
        try {
            return Integer.parseInt(s.trim());
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    private static boolean parseBool(String s, boolean fallback) {
        if (s == null || s.isEmpty()) return fallback;
        String low = s.trim().toLowerCase();
        return "true".equals(low) || "1".equals(low) || "yes".equals(low) || "y".equals(low);
    }
}
